Clean up message_ix.macro

behnam-zakeri opened the pull request:
This PR improves the code in `message_ix.macro.py` for the following items. Some of these issues were identified when #223 was conditionally merged, and some items are additional improvements.

- [x] testing zero or missing values in PRICE_COMMODITY _before_ calibration
- [x] removing hard-coded units in the code, and specifying them by user in the input data per requested in #317 
- [x] improving configuration for sets and specifying the "level" for macro outside the code (partly addressing #317)
- [x] mapping MACRO sectors to MESSAGE commodities (and level) explicitly by the user
- [x] removing old `initialize_year_macro` before calibration (partly addressing #320)
- [x] removing `baseyear_macro` from the code (not needed in GAMS anymore)
- [x] specifying MACRO years by user in the Excel "config" sheet, and slicing and reading data flexibly based on that.
- [x] adding tests and modifying the input data files to reflect the above changes 
- [x] adding docstring (addressing #315)

## How to review
New tests are added to check the improvements in this PR. You can view changes in the code and run the code and tests under `test_macro.py`.
Notice: if you wish to test the code on a input data file, please update your data for:
- "config": in the new format, there are three columns ("node, "sector", "level") in "config" to specify those members of these sets that are included in MACRO calibration. Please see the sample input file in: _message_ix/tests/data/westeros_macro_input.xlsx_
- you need to add a "unit" column to all input parameters for specifying the unit (Please see the format in the file above).

- [x] Tests added.
- [x] These improvements shouldn't change the main behaviour and documentation. However, this will change the format of the input data file. This has been explained in the documentation.
- [x] Release notes added, including the "migration notes" for the format of the input data file.

volker-krey commented:
I have some smaller suggestions to revise the macro_input.xlsx file. 

1. I would suggest to add year columns to price_ref, demand_ref and cost_ref parameters so that there is no ambiguity about which year the values provided refer to. 
2. In my view, it would be beneficial to explicitly define the years that MACRO is calibrated for. On the one hand, this would avoid mistakes with unintentionally added years in parameters like gdp_calibrate from which the years would need to be inferred otherwise. In addition, it allows flexibility to have more data in a file that is used (e.g., one could use the same file parameterizing a 10-year and a 5-year time step model version by just changing the year entries on the config tab).
3. The sector and commodity sets used by MACRO and MESSAGE don't have to use the same entries (although at present this is the case in all applications). I would therefore suggest to have an explicit mapping of (MACRO) sectors to (MESSAGE) commodity.level combinations (see example).

I have implemented the suggested modifications in the attached multiregion_macro_input.xlsx file with the changes marked in yellow.
[multiregion_macro_input.xlsx](https://github.com/iiasa/message_ix/files/4439075/multiregion_macro_input.xlsx)

behnam-zakeri commented:
Thanks @volker-krey for the suggestions and the sample file. I'll include them. I'll also try to see if we can make the calibration run for different lastmodel_years in the same model. That should also make it easier to calibrate a scenario with different last model year using the same input data file.

GamzeUnlu95 commented on message_ix/macro.py:
```diff
@@ -388,22 +402,24 @@ def add_model_data(base, clone, data):
-    for s in c.sectors:
-        clone.add_set("sector", s)
-        clone.add_set("mapping_macro_sector", [s, s, "useful"])
+    for s, l in product(c.sectors, c.levels):
```
I suggest modifying here to also differentiate the sectors and commodities which might not be always the same as suggested by Volker. 

```
for s, l, c in product(c.sectors, c.levels, c.commodities):
        clone.add_set('sector', s)
        clone.add_set("mapping_macro_sector", [s, c, l])
```

GamzeUnlu95 commented on message_ix/macro.py:
```diff
@@ -193,18 +190,28 @@ def __init__(self, s, data):
+        # users define certain nodes, sectors and level for MACRO
+        self.nodes = set(self.data['config']['node'].dropna())
+        self.sectors = set(self.data['config']['sector'].dropna())
+        self.levels = set(self.data['config']['level'].dropna())
```
Here we can also read the commodities from the configuration sheet to be later used in the "mapping_macro_sector"set:

`self.commodities = set(self.data['config']['commodity'].dropna())`

GamzeUnlu95 commented on message_ix/macro.py:
```diff
@@ -193,18 +190,28 @@ def __init__(self, s, data):
+        # users define certain nodes, sectors and level for MACRO
+        self.nodes = set(self.data['config']['node'].dropna())
+        self.sectors = set(self.data['config']['sector'].dropna())
+        self.levels = set(self.data['config']['level'].dropna())
```
I would suggest adding another line here to be able to obtain the unique mapping of the different sector, commodity, and level combinations. 

`self.sector_mapping = self.data["config"]`

behnam-zakeri replied:
Thanks @GamzeUnlu95 for the suggestion. Can you please make a pull request and suggest your changes?

GamzeUnlu95 replied:
Done in [this PR:](https://github.com/behnam-zakeri/message_ix/pull/9)

GamzeUnlu95 commented on message_ix/macro.py:
```diff
@@ -388,22 +402,24 @@ def add_model_data(base, clone, data):
-        clone.add_set("mapping_macro_sector", [s, s, "useful"])
+    for s, l in product(c.sectors, c.levels):
+        clone.add_set('sector', s)
+        clone.add_set("mapping_macro_sector", [s, s, l])
```
Here we assume sector and commodity names are the same. We could differentiate this by using a mapping table obtained from the excel "config" sheet including a commodity column. 

```
    sector_mapping_df = c.sector_mapping

    for i in sector_mapping_df.index:
        s = sector_mapping_df.iloc[i, sector_mapping_df.columns.get_loc("sector")]
        c = sector_mapping_df.iloc[i, sector_mapping_df.columns.get_loc("commodity")]
        l = sector_mapping_df.iloc[i, sector_mapping_df.columns.get_loc("level")]
        clone.add_set('sector', s)
        clone.add_set("mapping_macro_sector", [s, c, l])
```

behnam-zakeri replied:
@GamzeUnlu95 , nice suggestion. If possible, please make a PR and add these suggestions in the code, and in the Excel file, so that I can run and approve them.

GamzeUnlu95 replied:
Done in [this PR.](https://github.com/behnam-zakeri/message_ix/pull/9)

khaeru commented:
@behnam-zakeri once #484 is merged, you can rebase this, and the failure in the 'lint' check should disappear.

behnam-zakeri commented:
@volker-krey, we addressed your comments. I updated the description of the PR accordingly. I added docstring too and removed extra comments, to address #315. The only thing that in my view can be improved before merging is removing unit conversions from the code (at the moment, cost and price parameters are divided by 1000). This will also address #317. This may require removing these unit conversions from the GAMS code too. My question is:
1- can we remove the MACRO-related unit conversions from the GAMS code without any backward incompatibility? 
2- and on the practical point, should we do this through this PR, another related #430 by @GamzeUnlu95, or a new PR?

khaeru commented:
> The only thing that in my view can be improved before merging is removing unit conversions from the code (at the moment, cost and price parameters are divided by 1000).
> …
> 2. and on the practical point that @khaeru may instruct: should we do this through this PR, another related #430 by @GamzeUnlu95, or a new PR?

I would say another PR?

It's not necessary that this one PR do _all possible fixes and improvements_ to `.macro`. It's okay to leave that status quo behaviour in place (so long as it is documented!), and then use a separate PR to address it.

That approach also keeps this one smaller and easier to complete.

behnam-zakeri commented:
@khaeru , I rebased this PR and cleaned a bit. I run the tests without a problem on my machine (Windows) and the Windows checks pass here too. However, a few tests don't pass in Linux and Mac, which have the same cause. I feel this is caused by the way the data is being read from Excel. But I can't check this on my machine. Would you mind looking into this issue, or running `test_macro.py` of this PR on your machine?

khaeru commented:
> @khaeru Would you mind looking into this issue, or running `test_macro.py` of this PR on your machine?

I can do this after we wrap-up the 3.4 release, so sometime next week.

behnam-zakeri commented:
@LauWien I have a question. I get an lint error of `black` in the checks here, but when I run `black` over this file on my side the file remains "unchanged". I can't see what the issue is to resolve it in my code. How can I see that here from the lint check fail?

khaeru commented on message_ix/macro.py:
```diff
@@ -215,8 +259,14 @@ def read_data(self):
-                name, self.data[name], self.nodes, self.sectors, self.years
+                name, self.data[name], self.nodes, self.sectors, self.levels, self.years
             )
+            self.units[name] = self.data[name]["unit"].mode().any()
```
@behnam-zakeri here `any()` returns a boolean value (`True` or `False`), that is later inserted into some data frames, per the previous comment. Can you please clarify what the intent of this line is?

khaeru replied:
I am guessing it is intended to get a single unique unit value.

behnam-zakeri replied:
Yes, it's intended to get the unit that is used most, and use it as a single value.

khaeru commented on message_ix/macro.py:
```diff
@@ -438,13 +673,13 @@ def calibrate(s, check_convergence=True, **kwargs):
         .rename(columns={"lvl": "value"})
         .drop("mrg", axis=1)
-        .assign(unit=UNITS["aeei"])
+        .assign(unit=s.par("grow")["unit"].mode().any())
```
Same issue as in the other location.

khaeru commented:
Okay, I have hopefully fixed the unit errors mentioned in https://github.com/iiasa/message_ix/pull/327#pullrequestreview-911337274 and https://github.com/iiasa/message_ix/pull/327#pullrequestreview-911373638. I also applied latest black and rebased on `main`. Please `git pull --rebase` to retrieve the latest code.

behnam-zakeri commented:
Thanks @khaeru for resolving the technical issues and improving the code. I added documentation for the input data file, and also a few sentences as "Migration Notes" to warn that these changes will not work with the previous versions of the Excel data file.

khaeru commented:
Excellent. That looks like exactly what we need, so I will request a review and then @LauWien can dispatch that to an available colleague, e.g. as you suggested in Slack.

LauWien commented:
> @LauWien can dispatch that to an available colleague, e.g. as you suggested in Slack.

How soon should the PR get reviewed (E.g. best by the end of this week or end of next week)? Many thanks in advance!

khaeru commented:
End of next week is probably fine.

LauWien commented on message_ix/macro.py:
```diff
@@ -14,7 +15,6 @@
 
 You are using *experimental*, incomplete features from
 `message_ix.macro`—please exercise caution. Read more:
-- https://github.com/iiasa/message_ix/issues/315
```
Maybe we can still leave it in?

khaeru replied:
Yeah, agreed. Good catch!

LauWien commented on message_ix/macro.py:
```diff
@@ -37,15 +34,12 @@
+    price_MESSAGE="price_ref",
 )
 
 #: ixmp items (sets, parameters, variables, and equations) in MACRO.
```
```suggestion
# ixmp items (sets, parameters, variables, and equations) in MACRO.
```

LauWien commented on message_ix/macro.py:
```diff
@@ -384,26 +571,52 @@ def _aconst(self):
     c.read_data()
     c.derive_data()
 
+    # removing old initializeyear_macro before adding new one
```
```suggestion
    # Removing old initializeyear_macro before adding new one
```

LauWien commented on message_ix/macro.py:
```diff
@@ -384,26 +571,52 @@ def _aconst(self):
+    if not cat.empty:
+        clone.remove_set("cat_year", cat)
+
     # add temporal set structure
```
```suggestion
    # Add temporal set structure
```

LauWien commented on message_ix/macro.py:
```diff
@@ -384,26 +571,52 @@ def _aconst(self):
-    clone.add_set("type_year", "baseyear_macro")
-    clone.add_set("cat_year", ["baseyear_macro", c.base_year])
 
     # add nodal set structure
```
```suggestion
    # Add nodal set structure
```

LauWien commented on message_ix/macro.py:
```diff
@@ -384,26 +571,52 @@ def _aconst(self):
+        clone.add_set("sector", sec)
+        clone.add_set("mapping_macro_sector", [sec, com, lvl])
 
     # add parameters
```
```suggestion
    # Add parameters
```

LauWien commented on message_ix/macro.py:
```diff
@@ -412,18 +625,48 @@ def add_model_data(base, clone, data):
+        MACRO-calibrated scenario.
+
+    """
     # solve MACRO standalone
```
```suggestion
    # Solve MACRO standalone
```

LauWien commented on message_ix/macro.py:
```diff
@@ -433,18 +676,21 @@ def calibrate(s, check_convergence=True, **kwargs):
+    units = s.par("grow")["unit"].unique()
+    assert 1 == len(units), "Non-unique units for 'grow'"
+
     # get out calibrated values
```
```suggestion
    # Get out calibrated values
```

LauWien commented on message_ix/macro.py:
```diff
@@ -433,18 +676,21 @@ def calibrate(s, check_convergence=True, **kwargs):
+        .assign(unit=units[0])
     )
 
     # update calibrated value parameters
```
```suggestion
    # Update calibrated value parameters
```

LauWien commented on message_ix/tests/test_macro.py:
```diff
@@ -28,10 +28,10 @@ def var(self, name, **kwargs):
         df = self.data["aeei"]
         # add extra commodity to be removed
```
```suggestion
        # Add extra commodity to be removed
```

LauWien commented on message_ix/tests/test_macro.py:
```diff
@@ -28,10 +28,10 @@ def var(self, name, **kwargs):
         extra_commod = df[df.sector == "i_therm"].copy()
-        extra_commod["sector"] = self.data["config"]["ignore_sectors"][0]
+        extra_commod["sector"] = "bar"
         # add extra region to be removed
```
```suggestion
        # Add extra region to be removed
```

LauWien reviewed (approved):
Looks great :)
Just some suggestions.

khaeru commented:
@LauWien thanks for the review—please merge once checks again pass.
